### DISCUSSION TITLE
Litexcnc install

### DISCUSSION THREAD

**Discussion by horcsab** [Peter-van-Tol/LiteX-CNC#107] [General]
Dear Peter,

I replaced the PC with a new one and I am experiencing the following installation problem.
Linuxcnc Debian 12 Bookworm PREEMPT-RT ISO freshly installed.
Python3, pip3 installed and Litexcnc installed.

Oss-cad-suite is also installed.
When compiling firmware (litexcnc build_firmware *.json --build) I always get the error "Litex is not installed. Please run 'litexcnc install_litex' first."

If I issue the litexcnc command, it shows the litex commands (build_firmware, convert, flash,...).

What can I do wrong, what might not be set correctly?

**Comment by Peter-van-Tol** [comment]
At this moment I'm on holiday, so have this answer is from my mobile phone and from my head. My apologies, more detailed answer I can send on Sunday.

I see that the documentation / error message is maybe outdated. The command `litexcnc install_litex` has been replaced with `litexcnc install_toolchain`; did you run this command? If not, this will solve your error. 

Will change the error message.

**Reply by kzali** [reply]
@horcsab 

This is what I used on Ubuntu.  I  found this in one of the discussions:

sudo apt-get update && sudo apt-get -y upgrade
sudo apt-get install python3-pip python3-wheel python3-venv
python3 -m venv ~/.local --system-site-packages
echo -e "" >> ~/.bashrc
echo -e "# set PATH so it includes user's private bin if it exists" >> ~/.bashrc
echo -e "if [ -d \"$HOME/.local/bin\" ] ; then" >> ~/.bashrc
echo -e "    PATH=\"$HOME/.local/bin:$PATH\"" >> ~/.bashrc
echo -e "fi" >> ~/.bashrc
. ~/.bashrc
pip install litexcnc[cli]

litexcnc install_toolchain --user
sudo env "PATH=$PATH" litexcnc install_driver
. ~/.bashrc

Of course Peter is way more competent.  Maybe this helps

**Reply by Peter-van-Tol** [reply]
You're spot on with the recipe 👍, cannot improve on that.

**Reply by horcsab** [reply]
> @horcsab
> 
> This is what I used on Ubuntu. I found this in one of the discussions:
> 
> sudo apt-get update && sudo apt-get -y upgrade sudo apt-get install python3-pip python3-wheel python3-venv python3 -m venv ~/.local --system-site-packages echo -e "" >> ~/.bashrc echo -e "# set PATH so it includes user's private bin if it exists" >> ~/.bashrc echo -e "if [ -d "$HOME/.local/bin" ] ; then" >> ~/.bashrc echo -e " PATH="$HOME/.local/bin:$PATH"" >> ~/.bashrc echo -e "fi" >> ~/.bashrc . ~/.bashrc pip install litexcnc[cli]
> 
> litexcnc install_toolchain --user sudo env "PATH=$PATH" litexcnc install_driver . ~/.bashrc
> 
> Of course Peter is way more competent. Maybe this helps

kzali and Peter,
Thank you for your help!
The problem was solved, I reinstalled the machine and then installed it according to the above instructions. It worked!

Have a nice rest everyone!

Regards, Csaba

**Comment by horcsab** [comment]
So, next problem,

firmware build...

[FUW400-5A-75E.json](https://github.com/user-attachments/files/16637846/FUW400-5A-75E.json)

build error, 

Traceback (most recent call last):
  File "/home/cnc/.local/bin/litexcnc", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/home/cnc/.local/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cnc/.local/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/cnc/.local/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cnc/.local/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cnc/.local/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cnc/.local/lib/python3.11/site-packages/litexcnc/cli/build_firmware.py", line 37, in cli
    firmware_config = LitexCNC_Firmware.parse_raw(' '.join(config.readlines()))
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/main.py", line 549, in pydantic.main.BaseModel.parse_raw
  File "pydantic/main.py", line 526, in pydantic.main.BaseModel.parse_obj
  File "/home/cnc/.local/lib/python3.11/site-packages/litexcnc/firmware/soc.py", line 65, in __init__
    current_module = subclass(**current_module) 
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/main.py", line 341, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 4 validation errors for PWM_ModuleConfig
instances -> 0 -> pin
  field required (type=value_error.missing)
instances -> 1 -> pin
  field required (type=value_error.missing)
instances -> 2 -> pin
  field required (type=value_error.missing)
instances -> 3 -> pin
  field required (type=value_error.missing)

What am I not setting correctly in the pwm section?

**Reply by Peter-van-Tol** [reply]
Think that there are two possible issues:
1. The pwm direction is not yet released I think, so the version on PyPi (pip install) does not include this.
2. I have to check whether the keyword is pins or pin. You use pins, while the error says it expects pin (without the s)

You can change number 2 first to check, otherwise I help you with installing the latest version from git.

**Reply by horcsab** [reply]
I changed the keyword to pin, but the error is the same.

instances -> 0 -> pin
  str type expected (type=type_error.str)
instances -> 1 -> pin
  str type expected (type=type_error.str)
instances -> 2 -> pin
  str type expected (type=type_error.str)
instances -> 3 -> pin
  str type expected (type=type_error.str)

**Reply by Peter-van-Tol** [reply]
The error is different, it now says it expects the name of the pin. So this means you have to install the latest version from GitHub instead of PyPi.

`pip install git+https://github.com/Peter-van-Tol/LiteX-CNC`

**Reply by Peter-van-Tol** [reply]
Above should install the development version, which includes the feature you are looking for.

No further installations required.

**Reply by horcsab** [reply]
Thanks Peter!
It worked!

**Reply by Peter-van-Tol** [reply]
You're welcome.

Please share your experience with LitexCNC.

**Comment by horcsab** [comment]
Good morning,

nc@cnc:~$ halrun
halcmd: loadrt litexcnc connections="eth:192.168.12.20"
Note: Using POSIX realtime
litexcnc: Loading Litex CNC driver version 1.2.4
litexcnc: Loading and registering default modules:
litexcnc: Registered module gpio
litexcnc: Registered module pwm
litexcnc: Registered module encoder
litexcnc: Registered module step
litexcnc: Setting up board drivers: 
litexcnc: Registered driver eth
LitexCNC-eth: connected to board on '192.168.12.20:1234'
litexcnc: INFO: Version of firmware (1.3.0) is different with the version of the driver (1.2.4). Communication is still possible, although the firmware could use an update for the best experience. 
litexcnc: Setting up modules...
litexcnc: Reading 28 bytes
litexcnc:  - Watchdog
litexcnc:  - Wallclock
litexcnc:  - gpio ... done!
litexcnc: Out of memory!
litexcnc: Failed to instantiate module: 'pwm'
litexcnc:  - pwm ...board fails LitexCNC registration
litexcnc: Failed to initialize the driver.
litexcnc: rtapi_app_main: Invalid argument (-22)
<stdin>:1: waitpid failed /usr/bin/rtapi_app litexcnc
<stdin>:1: /usr/bin/rtapi_app exited without becoming ready
<stdin>:1: insmod for litexcnc failed, returned -1
halcmd: 


What could be the problem?
Flash, ping, ok.

**Comment by Peter-van-Tol** [comment]
I see this as a bug. Break down of what happens:
- the error only happens when memory cannot be allocated for a module, i.e.
`pwm` in this case;
- I don't think this is hardware related, so most-likely the byte with the
number of instances received from the FPGA is zero (which prevents
allocation of memory);
- either in the firmware there is an error or in the driver.

Going to check which part has been changed with the latest modification, as
this behavior I didn't experience before.

Op za 17 aug 2024 om 08:57 schreef Csaba Horvath ***@***.***>:

**Comment by Peter-van-Tol** [comment]
Quick check: after installing the development version of the firmware, did
you also re-install the driver? There has been a change in the driver which
could explain the behavior you are experiencing.

Op za 17 aug 2024 om 15:58 schreef Peter Van Tol ***@***.***>:

**Reply by horcsab** [reply]
A driver update solved this problem. 

However, I find that the ethernet connection is disconnected for some reason. During continuous pinging, the connection is interrupted more and more often, and then the connection with the card is terminated.

Could the card be faulty?
5A-75E v8.0

**Comment by horcsab** [comment]
I tried a previous firmware setting on the card and it works.
I am attaching the previous json file, this ok.
[CSABA-5A-75E.json](https://github.com/user-attachments/files/16645556/CSABA-5A-75E.json)


And new json, is not ok.
[FUW400-5A-75E.json](https://github.com/user-attachments/files/16645557/FUW400-5A-75E.json)

If you have some time, could you check what the error could be? I can't find it.
Thanks!

**Reply by horcsab** [reply]
I found the error. (tx_delay, txdelay.)

**Reply by Peter-van-Tol** [reply]
Does this mean that the connection is no longer dropping, or is this still an issue?

**Reply by horcsab** [reply]
The connection is also fine.

I will test the PWM_dir part soon. I made a PWM_dir to +-10V converter, which was fitted to an old servo drive.
It already works on the desktop, I want to test it with the machine next week. I will share my experience.

**Reply by Peter-van-Tol** [reply]
Great news! 

Could you also share your schematics of the converter? I think this is valuable for other users wanting so revive the old servos.

What kind of frequency do you run the PWM on? The commercial converters require PWM at a frequency of 1 kHz to 3 KHz. The resolution you get is then depending on the frequency of the FPGA. For example, when the card is running at 40 MHz, the maximum theoretical resolution is 40.000. This might not be enough when you want to run the servos in position mode.

When you run the servo in velocity mode, there might be a challenge in getting the feedback to LinuxCNC correct. Due to the communication with the FPGA, the command will be always a little later then with a solution which is completely done in the PC. I struggled a lot with this in designing the `stepgen` module.

So I'm really interested in your solution in this problem!

**Reply by Peter-van-Tol** [reply]
To add to above: If you need further precision, I can add dithering to the `pwm` module. This is currently not supported yet.

**Reply by horcsab** [reply]
Currently, 20kHz PWM is set for the test, but I will reduce this. I am attaching the existing documentation.
At the moment, I only have the basics, I still have a lot to improve on the settings, and the conversion electronics are not perfect yet.

I replaced the 74245 circuits on the colorlight board with SN74CBT3245APWR.With this, the level adjustment was solved and the outputs became bidirectional.

[+-10V_pdf_colorlight.pdf](https://github.com/user-attachments/files/16658546/%2B-10V_pdf_colorlight.pdf)

[fuw400.zip](https://github.com/user-attachments/files/16658488/fuw400.zip)

**Reply by Peter-van-Tol** [reply]
I like the schematics! Gives inspiration of creating my own 0-10 Volt converters.

As I can see in you your `HAL`, the servo's are running in position mode. Even when running as low as 1 kHz, I have the feeling that the resolution of 400.000 positions along an axis is too low. This ultimately depends on the size of the machine whether this is acceptable. Whilst having tinkered on PWM anyways, I think adding dithering is a good next step. The aim is to make PWM with full 32-bit resolution, independent on the frequency.

**Reply by Peter-van-Tol** [reply]
See also #108

**Reply by horcsab** [reply]
Sorry for my bad English. I will try to describe my experience.

I still don't understand a lot about linuxcnc, that's for sure.

The motor drive receives an analog signal, the range of which is +/-10V. According to my experience, it is sufficient to control this with a resolution of 1-2mV.

If I understand correctly, the positioning of linuxcnc works in such a way that the incremental encoder on the motor axis gives the position. And in linuxcnc there is the command position. If these two values ​​differ, the regulating +/-10V signal is created from the difference between these two values. The larger this error is, the larger the signal is, of course with reversed sign.

I don't understand why the resolution of 400,000 would be insufficient for PWM. A value of 40000 should be enough.
I just tried driving an old SOPREL DC motor on the table, which has a 500 pulse/revolution encoder at the end of the shaft.
The encoder is processed with 4x resolution by Linuxcnc and apparently works perfectly. Both at low and high speeds. The maximum speed of the motor is 3000rpm.

**Reply by Peter-van-Tol** [reply]
My bad, as I didn't read well enough your HAL-file. The output for your servo-drive determines the speed, not the position! In case of speed, the resolution is well enough. I can still learn everyday! With your explanation, I can follow what you are doing and it should work.

BTW: your English is good enough. It isn't my primary language either!

**Reply by AshGork** [reply]
> Currently, 20kHz PWM is set for the test, but I will reduce this. I am attaching the existing documentation. At the moment, I only have the basics, I still have a lot to improve on the settings, and the conversion electronics are not perfect yet.
> 
> I replaced the 74245 circuits on the colorlight board with SN74CBT3245APWR.With this, the level adjustment was solved and the outputs became bidirectional.
> 
> [+-10V_pdf_colorlight.pdf](https://github.com/user-attachments/files/16658546/%2B-10V_pdf_colorlight.pdf)
> 
> [fuw400.zip](https://github.com/user-attachments/files/16658488/fuw400.zip)

Hi horcsab,
I'm very intersted in your solution for the PWM (+dir) to +/-10V.
I don't understand why you have a 12/24V input on your schematics that powers the TLP118 ? And then connected to J3:7 ?

Thank you !

**Comment by EhsanRasouli1371** [comment]
hello
I did the installation steps as follows

```
sudo apt-get update && sudo apt-get -y upgrade
sudo apt-get install python3-pip python3-wheel python3-venv
python3 -m venv ~/.local --system-site-packages
echo -e "" >> ~/.bashrc
echo -e "# set PATH so it includes user's private bin if it exists" >> ~/.bashrc
echo -e "if [ -d "$HOME/.local/bin" ] ; then" >> ~/.bashrc
echo -e " PATH="$HOME/.local/bin:$PATH"" >> ~/.bashrc
echo -e "fi" >> ~/.bashrc
. ~/.bashrc
pip install litexcnc[cli]

litexcnc install_toolchain --user
sudo env "PATH=$PATH" litexcnc install_driver
. ~/.bashrc
```

But I get this error when building the firmware
[config.json](https://github.com/user-attachments/files/18244658/config.json)


```
3.43. Executing ABC pass (technology mapping using ABC).

3.43.1. Extracting gate netlist of module `\colorlight_5a_75e' to `<abc-temp-dir>/input.blif'..
Extracted 19030 gates and 26137 wires to a netlist network with 7105 inputs and 4306 outputs.

3.43.1.1. Executing ABC.
Running ABC command: "<yosys-exe-dir>/yosys-abc" -s -f <abc-temp-dir>/abc.script 2>&1
ABC: Bus error
ERROR: ABC: execution of command ""/home/ehsan/toolchain/oss-cad-suite/lib/yosys-abc" -s -f /tmp/yosys-abc-Uy7EKr/abc.script 2>&1" failed: return code 135.
Traceback (most recent call last):
  File "/home/ehsan/.local/bin/litexcnc", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/home/ehsan/.local/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ehsan/.local/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/ehsan/.local/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ehsan/.local/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ehsan/.local/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ehsan/.local/lib/python3.11/site-packages/litexcnc/cli/build_firmware.py", line 58, in cli
    builder.build(run=build)
  File "/home/ehsan/toolchain/litex/litex/litex/soc/integration/builder.py", line 350, in build
    vns = self.soc.build(build_dir=self.gateware_dir, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ehsan/toolchain/litex/litex/litex/soc/integration/soc.py", line 1205, in build
    return self.platform.build(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ehsan/toolchain/litex/litex/litex/build/lattice/platform.py", line 43, in build
    return self.toolchain.build(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ehsan/toolchain/litex/litex/litex/build/lattice/trellis.py", line 236, in build
    _run_script(script)
  File "/home/ehsan/toolchain/litex/litex/litex/build/lattice/trellis.py", line 174, in _run_script
    raise OSError("Error occured during Yosys/Nextpnr's script execution.")
OSError: Error occured during Yosys/Nextpnr's script execution.
```

Can someone tell me how to fix this error?

**Reply by Peter-van-Tol** [reply]
Will help you after Christmas. Wish you all nice days with family and friends!

**Reply by EhsanRasouli1371** [reply]
Thank you dear Peter
I also wish you and your esteemed family a year full of success

**Comment by Peter-van-Tol** [comment]
I would like to recommend to install a newer version of LitexCNC, by using the following command:
``` bash
pip install git+https://github.com/Peter-van-Tol/LiteX-CNC@119-feature-request-revamp-watchdog
```

In this version, a few bugs have been resolved and new features have been implemented:
- better watchdog (#119)
- user LED and button supported (#117)
- more stepgen modes supported (#104), also fixed bugs in the stepgen, making stepping at higher rates possible
- different PWM modes supported (#104)
- fixed version of the toolchain

To use these version, it is required to re-install both the driver and toolchain:
```
litexcnc install_toolchain --user
litexcnc install_driver
```

See attached file with the required changes.

[config.json](https://github.com/user-attachments/files/18304399/config.json)

If you want, I can also share the build firmware.

**Reply by EhsanRasouli1371** [reply]
grateful

**Reply by EhsanRasouli1371** [reply]
How do I fix this error?
```
$ pip install git+https://github.com/Peter-van-Tol/LiteX-CNC@119-feature-request-revamp-watchdog
Collecting git+https://github.com/Peter-van-Tol/LiteX-CNC@119-feature-request-revamp-watchdog
  Cloning https://github.com/Peter-van-Tol/LiteX-CNC (to revision 119-feature-request-revamp-watchdog) to /tmp/pip-req-build-jfnt3ggu
  Running command git clone --filter=blob:none --quiet https://github.com/Peter-van-Tol/LiteX-CNC /tmp/pip-req-build-jfnt3ggu
  error: RPC failed; curl 92 HTTP/2 stream 3 was not closed cleanly: CANCEL (err 8)
  error: 4547 bytes of body are still expected
  fetch-pack: unexpected disconnect while reading sideband packet
  fatal: early EOF
  fatal: fetch-pack: invalid index-pack output
  fatal: could not fetch 72b463fc04c20757b1b565ea93c867a2f2dc84b2 from promisor remote
  warning: Clone succeeded, but checkout failed.
  You can inspect what was checked out with 'git status'
  and retry with 'git restore --source=HEAD :/'

  error: subprocess-exited-with-error
  
  × git clone --filter=blob:none --quiet https://github.com/Peter-van-Tol/LiteX-CNC /tmp/pip-req-build-jfnt3ggu did not run successfully.
  │ exit code: 128
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× git clone --filter=blob:none --quiet https://github.com/Peter-van-Tol/LiteX-CNC /tmp/pip-req-build-jfnt3ggu did not run successfully.
│ exit code: 128
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```

**Reply by EhsanRasouli1371** [reply]
Please also share the build firmware of this file
[config.json](https://github.com/user-attachments/files/18307887/config.json)
Thanks

**Reply by Peter-van-Tol** [reply]
For the error: I think your internet connection was intermittent or something on your PC is not correctly configured. When trying the command, it installs just fine:
```
pip install git+https://github.com/Peter-van-Tol/LiteX-CNC@119-feature-request-revamp-watchdog
OUTPUT:
  Cloning https://github.com/Peter-van-Tol/LiteX-CNC (to revision 119-feature-request-revamp-watchdog) to c:\users\tolp2\appdata\local\temp\pip-req-build-ipf2_o95
  Running command git clone --filter=blob:none --quiet https://github.com/Peter-van-Tol/LiteX-CNC 'C:\Users\tolp2\AppData\Local\Temp\pip-req-build-ipf2_o95'
  Running command git checkout -b 119-feature-request-revamp-watchdog --track origin/119-feature-request-revamp-watchdog
  branch '119-feature-request-revamp-watchdog' set up to track 'origin/119-feature-request-revamp-watchdog'.
  Switched to a new branch '119-feature-request-revamp-watchdog'
  Resolved https://github.com/Peter-van-Tol/LiteX-CNC to commit 9e7386738cd8641af80cb5f943daf46096050465
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting chardet<6.0.0,>=5.0.0 (from litexcnc==1.3.0.dev2)
  Using cached chardet-5.2.0-py3-none-any.whl.metadata (3.4 kB)
Collecting click<9.0.0,>=8.1.3 (from litexcnc==1.3.0.dev2)
  Downloading click-8.1.8-py3-none-any.whl.metadata (2.3 kB)
Collecting meson<0.65.0,>=0.64.1 (from litexcnc==1.3.0.dev2)
  Using cached meson-0.64.1-py3-none-any.whl.metadata (1.6 kB)
Collecting ninja<2.0.0,>=1.11.1 (from litexcnc==1.3.0.dev2)
  Downloading ninja-1.11.1.3-py3-none-win_amd64.whl.metadata (5.3 kB)
Collecting packaging<22.0,>=21.3 (from litexcnc==1.3.0.dev2)
  Using cached packaging-21.3-py3-none-any.whl.metadata (15 kB)
Collecting pydantic<2.0.0,>=1.10.2 (from litexcnc==1.3.0.dev2)
  Downloading pydantic-1.10.19-py3-none-any.whl.metadata (152 kB)
Collecting requests<3.0.0,>=2.28.1 (from litexcnc==1.3.0.dev2)
  Using cached requests-2.32.3-py3-none-any.whl.metadata (4.6 kB)
Collecting yapps<3.0.0,>=2.2.0 (from litexcnc==1.3.0.dev2)
  Using cached Yapps-2.2.0.tar.gz (13 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting colorama (from click<9.0.0,>=8.1.3->litexcnc==1.3.0.dev2)
  Using cached colorama-0.4.6-py2.py3-none-any.whl.metadata (17 kB)
Collecting pyparsing!=3.0.5,>=2.0.2 (from packaging<22.0,>=21.3->litexcnc==1.3.0.dev2)
  Downloading pyparsing-3.2.1-py3-none-any.whl.metadata (5.0 kB)
Collecting typing-extensions>=4.2.0 (from pydantic<2.0.0,>=1.10.2->litexcnc==1.3.0.dev2)
  Using cached typing_extensions-4.12.2-py3-none-any.whl.metadata (3.0 kB)
Collecting charset-normalizer<4,>=2 (from requests<3.0.0,>=2.28.1->litexcnc==1.3.0.dev2)
  Downloading charset_normalizer-3.4.1-cp313-cp313-win_amd64.whl.metadata (36 kB)
Collecting idna<4,>=2.5 (from requests<3.0.0,>=2.28.1->litexcnc==1.3.0.dev2)
  Using cached idna-3.10-py3-none-any.whl.metadata (10 kB)
Collecting urllib3<3,>=1.21.1 (from requests<3.0.0,>=2.28.1->litexcnc==1.3.0.dev2)
  Downloading urllib3-2.3.0-py3-none-any.whl.metadata (6.5 kB)
Collecting certifi>=2017.4.17 (from requests<3.0.0,>=2.28.1->litexcnc==1.3.0.dev2)
  Downloading certifi-2024.12.14-py3-none-any.whl.metadata (2.3 kB)
Using cached chardet-5.2.0-py3-none-any.whl (199 kB)
Downloading click-8.1.8-py3-none-any.whl (98 kB)
Using cached meson-0.64.1-py3-none-any.whl (895 kB)
Downloading ninja-1.11.1.3-py3-none-win_amd64.whl (296 kB)
Using cached packaging-21.3-py3-none-any.whl (40 kB)
Downloading pydantic-1.10.19-py3-none-any.whl (165 kB)
Using cached requests-2.32.3-py3-none-any.whl (64 kB)
Downloading certifi-2024.12.14-py3-none-any.whl (164 kB)
Downloading charset_normalizer-3.4.1-cp313-cp313-win_amd64.whl (102 kB)
Using cached idna-3.10-py3-none-any.whl (70 kB)
Downloading pyparsing-3.2.1-py3-none-any.whl (107 kB)
Using cached typing_extensions-4.12.2-py3-none-any.whl (37 kB)
Downloading urllib3-2.3.0-py3-none-any.whl (128 kB)
Using cached colorama-0.4.6-py2.py3-none-any.whl (25 kB)
Building wheels for collected packages: litexcnc, yapps
  Building wheel for litexcnc (pyproject.toml) ... done
  Created wheel for litexcnc: filename=litexcnc-1.3.0.dev2-py3-none-any.whl size=153429 sha256=2dc08dcbb3b3ade68f1f6d56d892fa37629a2b7b9aef588e72ffdf859d499513
  Stored in directory: C:\Users\tolp2\AppData\Local\Temp\pip-ephem-wheel-cache-0xpa9y3b\wheels\f8\89\cc\a34ac3a444618023cda0e659bd0af3492bb6e473a47376ff92
  Building wheel for yapps (pyproject.toml) ... done
  Created wheel for yapps: filename=Yapps-2.2.0-py3-none-any.whl size=15824 sha256=a311aa713a6ebefa1638f80da3ef06a4693f859fccbc93636f2d836dd3059099
  Stored in directory: c:\users\tolp2\appdata\local\pip\cache\wheels\62\35\b8\d8dacd0f138f0cf12633eeb3708fd026427c0b2eaab8338538
Successfully built litexcnc yapps
Installing collected packages: yapps, urllib3, typing-extensions, pyparsing, ninja, meson, idna, colorama, charset-normalizer, chardet, certifi, requests, pydantic, packaging, click, litexcnc
Successfully installed certifi-2024.12.14 chardet-5.2.0 charset-normalizer-3.4.1 click-8.1.8 colorama-0.4.6 idna-3.10 litexcnc-1.3.0.dev2 meson-0.64.1 ninja-1.11.1.3 packaging-21.3 pydantic-1.10.19 pyparsing-3.2.1 requests-2.32.3 typing-extensions-4.12.2 urllib3-2.3.0 yapps-2.2.0
```

Could you try again. And if it does not work, try to get some more information on the error with `git status` as suggested in the error.

**Comment by romanetz** [comment]
Well, I had the same issue. Litexcnc, been installed from pip repository,
creates script named litexcnc in virtualenv/bin directory. Litexcnc, been
installed from git repository, doesn't create the script but is callable in
virtualenv via call: python -m litexcnc

вс, 5 янв. 2025 г. в 00:27, Peter-van-Tol ***@***.***>:

**Reply by EhsanRasouli1371** [reply]
How do I fix this error?

```
:~$ python -m litexcnc install_driver
INFO: Looking for halcompile...
INFO: Locating Makefile.modinc...
INFO: Run In Place (RIP) Installation not detected
INFO: Detected EMC2_RTLIB_DIR location '/usr/lib/linuxcnc/modules'
INFO: Retrieving default driver files to compile...
Copying file 'litexcnc.c'
Copying file 'watchdog.c'
Copying file 'wallclock.c'
Copying file 'litexcnc_pwm.c'
Copying file 'litexcnc_gpio.c'
Copying file 'litexcnc_stepgen.c'
Copying file 'litexcnc_encoder.c'
Copying file 'litexcnc_pigpio_speed_test.c'
Copying file 'etherbone.c'
Copying file 'litexcnc_eth.c'
Copying file 'litexcnc_spidev.c'
Copying file 'litexcnc_pigpio.c'
Copying file 'litexcnc.h'
Copying file 'wallclock.h'
Copying file 'watchdog.h'
Copying file 'litexcnc_gpio.h'
Copying file 'litexcnc_stepgen.h'
Copying file 'litexcnc_encoder.h'
Copying file 'litexcnc_pwm.h'
Copying file 'litexcnc_spidev.h'
Copying file 'litexcnc_pigpio.h'
Copying file 'etherbone.h'
Copying file 'litexcnc_eth.h'
INFO: Compiling LitexCNC driver...
Compiling realtime litexcnc.c
Linking litexcnc.so
cp litexcnc.so /usr/lib/linuxcnc/modules/
cp: cannot create regular file '/usr/lib/linuxcnc/modules/litexcnc.so': Permission denied
make: *** [/usr/share/linuxcnc/Makefile.modinc:107: install] Error 1
Error: Compilation of the driver failed.

```

**Reply by Peter-van-Tol** [reply]
To install the driver, you need `sudo` to run. However, when running `sudo`, it will not take the correct environment.

You can use the following command, to keep the current PATH variables untouched:
```bash
sudo -E env PATH=$PATH litexcnc install_driver
```
or, when `litexcnc` is not recognized:
```bash
sudo -E env PATH=$PATH python -m litexcnc install_driver
```

I thought it was in the documentation, but this was not released (yet). My bad...

**Reply by Peter-van-Tol** [reply]
Version 1.3.0. has just been released. Also, the documentation has been revised!

**Reply by EhsanRasouli1371** [reply]
🙏🙏🙏

**Comment by romanetz** [comment]
Just change permissions of the destination folder, make it writable to
regular user: sudo chmod 777 /usr/lib/linuxcnc/modules Also, of course, a
recompilation and reflashing of the board is needed.

вс, 5 янв. 2025 г., 14:05 EhsanRasouli1371 ***@***.***>:

**Reply by Peter-van-Tol** [reply]
Thanks for your support helping other people to use this project.

**Comment by EhsanRasouli1371** [comment]
Hi Peter
I installed Litexcnc version 1.3.3. Apparently there is a problem with the direction pin,
the motor direction does not change.

**Reply by Peter-van-Tol** [reply]
I know. Has been solved yesterday.

Please install version 1.3.4 and recompile your firmware. Also reinstall the driver if you didn't do it already, as this was required for version 1.3.3.

**Comment by horcsab** [comment]
Hello Peter,

I have a problem, please help me.
I upgraded the firmware from 1.3.0 to 1.3.4. With the halrun show command I can't see the PWM_freq and some gpio inputs and outputs. The firmware build ran successfully, I loaded it into the card and then rebooted the card and I experience this.
I am attaching the json file.

I burned it with an FTDI programmer with the following command:
 litexcnc flash_firmware --programmer ft232r --permanent colorlight_5a_75e.svf

What could be the problem?

[FUW400-5A-75E_watchdog.json](https://github.com/user-attachments/files/20126423/FUW400-5A-75E_watchdog.json)

**Comment by EhsanRasouli1371** [comment]
I installed litexcnc on my Raspberry Pi 4
I keep getting a watchdog error.
How do I fix this error?
SERVO_PERIOD=1000000
watchdog.timeout-ns 1500000

**Reply by Peter-van-Tol** [reply]
When do you get the watchdog error?

If it is during operation, you might have to check the connections, especially when using SPI. If it is after / during closing of halrun or LinuxCNC, this is not harmful.

Can you share the complete HAL?

**Reply by EhsanRasouli1371** [reply]
This error occurs during operation.
I am using a network connection and have checked the connections.
I also used a shielded cable.

[hal.txt](https://github.com/user-attachments/files/22669509/hal.txt)
[ini.txt](https://github.com/user-attachments/files/22669514/ini.txt)

**Reply by Peter-van-Tol** [reply]
Can you share the error?

When you start LinuxCNC from the command line, it will print out a lot of debug information. Also some information about the status of the watchdog.

**Reply by EhsanRasouli1371** [reply]
![IMG_20251003_004201_021](https://github.com/user-attachments/assets/71018aa0-5eb9-4a69-88eb-d0790b6094d0)

**Reply by EhsanRasouli1371** [reply]
It seems that this problem is either due to the speed of the Raspberry Pi's processor or its network card.
Because when I use a PC this problem doesn't exist.

**Reply by Peter-van-Tol** [reply]
Thank you for the image. This error occurs when the time-out happened at the FPGA. To be sure:
- what is the firmware version;
- what is your driver version.

Just to rule out that the versions are not aligned and the wrong value has been read. Can you also send your json? 

Will look tomorrow further...

**Reply by EhsanRasouli1371** [reply]
I installed Litexcnc version 1.3.4, but I don't know how to find the firmware and driver version.
[config.json](https://github.com/user-attachments/files/22676074/config.json)
